### PR TITLE
Reconcile Schema Comparison

### PR DESCRIPTION
**Pull request by ganeshdogiparthi-db**
* Using sqlglot to parse the source datatype and compare if it is an exact match against the target(databricks) type.
* rowid, urowid, anytype, anydata and anydataset sqlglot doesn't have these tokens. So, these datatypes in oracle are not handled.

### CODE REVIEW

File: src/databricks/labs/remorph/reconcile/comparison/schema.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import logging
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/comparison/schema.py:1]: file can be in base `reconcile` package

File: src/databricks/labs/remorph/reconcile/constants.py [hunk 2]
```diff
@@ -87,6 +89,17 @@ class Constants:
         },
     }
 
+    # Define the schema for the schema compare DataFrame
+    schema_compare = StructType(
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/constants.py:93]: Move this to schema compare file as private variable

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    SchemaMatchResult,
+    SchemCompareOutput,
+    Table,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SchemaCompare:
+    def __init__(
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:21]:
```suggestion
    def __init__( self, table_conf, spark) 
```
Move the rest of the parameters to the individual methods.

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            f"""
+        Snowflake datatype: create table dummy ({master.source_column} {master.source_datatype})
+        Parse datatype: {parsed_query}
+        Databricks datatype: {databricks_query}
+        """
+        )
+        if parsed_query.lower() != databricks_query.lower():
+            master.is_valid = False
+
+    def compare(self) -> SchemCompareOutput:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:108]: this method should take 

 ```suggestion
    def compare(self,         
       source_schema: list[Schema],
        databricks_schema: list[Schema],
        source: str
        ) -> SchemCompareOutput:
```

File: tests/unit/reconcile/conftest.py [hunk 5]
```diff
@@ -105,3 +105,157 @@ def schema():
 @pytest.fixture
 def expr():
     return parse_one("SELECT col1 FROM DUAL")
+
+
+@pytest.fixture
+def snowflake_databricks_schema():
```
Comment (sundarshankar89) [inline tests/unit/reconcile/conftest.py:111]: move to the individual test_schema_compare file only.

Comment (sundarshankar89) [reply tests/unit/reconcile/conftest.py:111]: and also you can simplify this using a small function

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/schema_compare.py (hunk 4 above)
Comment (ganeshdogiparthi-db) [reply src/databricks/labs/remorph/reconcile/schema_compare.py]: Done.

File: tests/unit/reconcile/conftest.py (hunk 5 above)
Comment (ganeshdogiparthi-db) [reply tests/unit/reconcile/conftest.py]: Done. Following the same approach for now.

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        :return: DataFrame
+        """
+        data = [tuple(asdict(item).values()) for item in data]
+        df = self.spark.createDataFrame(data, schema)
+
+        return df
+
+    @classmethod
+    def _parse(cls, source: str, column: str, data_type: str) -> str:
+        config = MorphConfig(source=source)
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:83]: Don't use this
```suggestion
```

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        return (
+            parse_one(f"create table dummy ({column} {data_type})", read=config.get_read_dialect())
+            .sql(dialect=config.get_write_dialect())
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:86]:
```suggestion
            .sql(dialect=SQLGLOT_DIALECTS.get("databricks"))
```

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+            .replace(", ", ",")
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:87]: why do you need this replace?

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        )
+
+    @classmethod
+    def _all_valid(cls, schema_compare_maps: list[SchemaMatchResult]) -> bool:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:91]:
```suggestion
    def _table_schema_status(cls, schema_compare_maps: list[SchemaMatchResult]) -> bool:
```

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+        return bool(all(x.is_valid for x in schema_compare_maps))
+
+    @classmethod
+    def _validate_parsed_query(cls, master: SchemaMatchResult, parsed_query) -> None:
+        databricks_query = f"create table dummy ({master.source_column} {master.databricks_datatype})"
+        logger.info(
+            f"""
+        Snowflake datatype: create table dummy ({master.source_column} {master.source_datatype})
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:99]: Check this log message again, why do we have Snowflake here?

File: src/databricks/labs/remorph/reconcile/schema_compare.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        data = [tuple(asdict(item).values()) for item in data]
+        df = self.spark.createDataFrame(data, schema)
+
+        return df
+
+    @classmethod
+    def _parse(cls, source: str, column: str, data_type: str) -> str:
+        config = MorphConfig(source=source)
+        return (
+            parse_one(f"create table dummy ({column} {data_type})", read=config.get_read_dialect())
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/schema_compare.py:85]:
```suggestion
            parse_one(f"create table dummy ({column} {data_type})", read=SQLGLOT_DIALECTS.get(source))
```

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/schema_compare.py (hunk 8 above)
Comment (ganeshdogiparthi-db) [reply src/databricks/labs/remorph/reconcile/schema_compare.py:87]: Parsing adds space, suppose like DECIMAL(10, 20), but when you consume schema from information schema/describe command, the output is like DECIMAL(10,20) .. to handle that, I am doing this.

**Review by sundarshankar89 — APPROVED** [review]
LGTM

**Pull request merged**